### ISSUE TITLE
Discussion: interop with closure compiler

### ISSUE THREAD

**Issue by LStAmour** [bazelbuild/rules_nodejs#966]
Note: This ticket is actually about external use of the examples, or how you'd use these rules in practice. But if you get technical, the bug to fix is that examples are tested only with the internal compiler, and so you have to override the compiler to run the examples for closure interop.

I was trying to get `tsickle` to run and produce output suitable for use with `rules_closure`. I had my `tsconfig.json` set up as:

```json
{
    "compilerOptions": {
        "target": "es5",
        "types": []
    },
    "bazelOptions": {
        "es5Mode": true,
        "tsickle": true,
        "tsickleGenerateExterns": true
    }
}
```

I was trying to get my `background.ts` file to build using tsickle and an example's es5_consumer:

```python
load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")

ts_library(
    name = "background",
    srcs = ["background.ts"],
    tsconfig = "tsconfig.json",
)

load("@build_bazel_rules_typescript//examples/es5_output:es5_consumer.bzl", "es5_consumer")

es5_consumer(
    name = "background_es5",
    deps = [":background"],
)
```

But I kept getting the following error:

```
Error: When setting bazelOpts { tsickle: true }, you must also add a devDependency on the tsickle npm package
    at emitWithTsickle (.../bazel-out/host/bin/external/build_bazel_rules_typescript/@bazel/typescript/tsc_wrapped.runfiles/npm/node_modules/@bazel/typescript/tsc_wrapped/tsc_wrapped.js:290:19)
```

And nothing I did seemed to resolve it.

Eventually I started digging into `rules_typescript`, and discovered the custom internal compiler still had a dependency on `@npm//tsickle` so I managed to write my own compiler wrapper to fix this:

```python
load("@build_bazel_rules_nodejs//:defs.bzl", "nodejs_binary")

# The following is required to run tsickle
nodejs_binary(
    name = "tsc_wrapped",
    entry_point = "@bazel/typescript/tsc_wrapped/tsc_wrapped.js",
    # The --expose-gc node option is required for tsc_wrapped
    templated_args = ["--node_options=--expose-gc"],
    data = [
        "@npm//protobufjs",
        "@npm//source-map-support",
        "@npm//tsickle",
        "@npm//tsutils",
        "@npm//typescript",
        "@npm//@bazel/typescript",
    ]
)

load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")

ts_library(
    name = "background",
    compiler = ":tsc_wrapped",
    srcs = ["background.ts"],
    tsconfig = "tsconfig.json",
)

load("@build_bazel_rules_typescript//examples/es5_output:es5_consumer.bzl", "es5_consumer")

es5_consumer(
    name = "background_es5",
    deps = [":background"],
)
```

... and everything started working as expected! 👍 

So I'm filing this issue for two reasons:

1. It would be nice if there were a way to run the examples using the external, public compiler instead of the internal one. (Or to enforce similar enough behaviour from both compilers...)

2. Any suggestions on how to better run `rules_typescript`, `tsickle` and `rules_closure` together? Am I on the right track? I'd appreciate any pointers or tips, I'm still very new to Bazel, but I'm looking to use the public APIs rather than relying on third-party forks for this integration.

**Comment by alexeagle** [comment]
`ts_library` is probably going to be deprecated, but you can continue using the `@bazel/concatjs` npm package which contains it anyway.

If you can express your transformations purely in terms of the CLI of tools, such as tsickle and google-closure-compiler, you could in theory do this in Bazel with just a macro that wraps the generated `bin` entries for those tools. However in my experience Closure is super complicated to interop with anything. https://github.com/bazelbuild/rules_closure might offer solutions but hasn't been released in a year.

Sorry, I think there's just not enough adoption of these tools outside of Google for anyone to make a living supporting them, they're not much fun to hobby on, and Google themselves don't put much effort behind making them usable externally...

**Comment by jimmyt857** [comment]
Yeah that makes sense, thank you for the quick response. It's a shame there isn't more support here, as typescript + closure seem like a perfect fit to me: if there was a seamless integration between them, then everyone could have the best of both worlds (cleanliness of typescript, optimizations of closure).

I'll see what I can do, but will probably ending up sticking with what we have for now then.
